feat: add instructions in case the release fails (#3433)
facebook/react-native-website: website/contributing/release-troubleshooting.md

title: Troubleshooting
---

### Adding a Local Commit in a -stable branch

> In the `-stable` branches (contrary to main branch), the release crew can directly push new commits to quickly address issues. After applying the necessary changes, those changes need to be re-implemented in `main`; this is to avoid inconsistent state across branches, and to prevent having the same failures in the future releases.

Reviewer: ```suggestion
In the `-stable` branches (contrary to main branch), the release crew can directly push new commits to quickly address issues. After applying the necessary changes, those changes need to be re-implemented in `main`; this is to avoid inconsistent state across branches, and to prevent having the same failures in future releases.
```

> To make sure not to forget about these changes, we follow some best practices.

Reviewer: ```suggestion
To make sure not to forget about these changes, we follow some best practices:
```

In case Step 1 completes successfully but Step 2 fails, the code in the stable branch is set to the new version, but it has failed to rollout. This means that we need to do some extra steps to be able to retry to publish this version.
If you end up in this situation, follow these steps:

1. Remove the tag that has been pushed (update the `X`, `Y` and `k` values appropriately):

```sh
git tag -d v0.X.Y[-RC.k]
git push origin :v0.X.Y[-RC.k]
```

1. Revert the commit pushed by Step 1. The script we use for the release updates the version numbers all across the codebase in many files ([here's an example](https://github.com/facebook/react-native/commit/38465f2d184a2558681a6b6b45163694c885bd39)). The commit usually has this title `[0.X.Y-rc.k] Bump version numbers` for RCs or `[0.X.Y] Bump version numbers` for stable releases.

```sh
git revert <sha of the commit>
git push origin 0.X-stable
```

1. Apply all the necessary fixes in the `-stable` branch of the codebase. The remedies for this could be of two types:
   - you forgot to cherry-pick some commits. In this case, you follow the standard cherry-pick procedure for RCs and patches.

>    - you need to fix some bugs in the code. See the ["Add a local commit"](#adding-a-local-commit-in-a--stable-branch) section below.

Reviewer: ```suggestion
   - you need to fix some bugs in the code. See the ["Add a local commit"](#adding-a-local-commit-in-a--stable-branch) section above.
```

> 1. Once you have pushed all the necessary commits in the branch, re-trigger the release process **with the same tag** via the usual script.

Reviewer: in the preview deployment it looks like markdown is not able to keep the numeration: https://deploy-preview-3433--react-native.netlify.app/contributing/release-troubleshooting#the-git-tag-has-been-pushed-to-remote-but-ci-failed-to-release-to-npm

can you hardcode the steps numbers?

# Make sure you push any local tags you added/removed
git push <remote> <branch> --follow-tags
```

> ### The Tag Has Been Pushed but NPM Failed to Publish React Native

Reviewer: first feedback: can you put these two sections as top ones, with the `Add a Local Commit` one first, then this, then the other two (I think it'd be better to have them in order of frequency, from the most likely one to the least)

Reviewer: also: 
```suggestion
### The Git Tag Has Been Pushed to Remote but CI Failed to Release to NPM
```

> The release process takes various steps, but they can be summarised with these two:

Reviewer: ```suggestion
The release process is composed of various steps, of which two are key for the rollout:
```

1. Create a version tag on the GitHub repository
2. Publish all the artifacts on Maven and NPM

> Step 2 takes some time because it has to build everything we need to publish and it can happen that some builds fail or there are other issues.

Reviewer: ```suggestion
Step 2 takes most of the CI time because it has to build all the artifacts and prefabs needed - so, it might happen that CI will fail during this part, hence not reaching the publish to npm step.
```

> In case Step 1 completes successfully but step 2 fails, the repo is in a dirty situation and you need to address that.

Reviewer: ```suggestion
In case Step 1 completes successfully but Step 2 fails, the code in the stable branch is set to the new version, but it has failed to rollout. This means that we need to do some extra steps to be able to retry to publish this version.
```

If you end up in this situation, follow these steps:

1. Remove the tag that has been pushed (update the `X`, `Y` and `k` values appropriately):

```sh
git tag -d v0.X.Y[-RC.k]
git push origin :v0.X.Y[-RC.k]
```

> 1. Revert the commit pushed by Step 1. The script we use for the release updates some versions in the `package.json`. The commit usually has this title `[0.X.Y-rc.k] Bump version numbers` for RCs or `[0.X.Y] Bump version numbers` for stable releases.

Reviewer: ```suggestion
1. Revert the commit pushed by Step 1. The script we use for the release updates the version numbers all across the codebase in many files ([here's an example](https://github.com/facebook/react-native/commit/38465f2d184a2558681a6b6b45163694c885bd39)). The commit usually has this title `[0.X.Y-rc.k] Bump version numbers` for RCs or `[0.X.Y] Bump version numbers` for stable releases.
```

```sh
git revert <sha of the commit>
git push origin 0.X-stable
```

> 1. Apply all the changes to fix the broken release. The remedies for this can could be of two types:

Reviewer: ```suggestion
1. Apply all the necessary fixes in the -stable branch of the codebase. The remedies for this can could be of two types:
```

   - you forgot to cherry-pick some commits. In this case, you follow the standard cherry-pick procedure for RCs and patches.
   - you need to fix some bugs in the code. See the ["Add a local commit"](#add-a-local-commit) section below.

> 1. Restart the release process **with the same tag**

Reviewer: ```suggestion
1. Once you have pushed all the necessary commits in the branch, re-trigger the release process **with the same tag** via the usual bump-oss script
```

Author: I tried not to explicitly mention the `bump-oss` script to be future-proof. Imagine that in a year we remove that script for something else, we would have to update this doc page as well.

> ### Add a Local Commit

Reviewer: ```suggestion
### Adding a Local Commit in a -stable branch
```

> After applying some changes in the repo to fix the release issues, you have to report those changes back in `main`. Otherwise, you risk to have the same failures in the next release.

Reviewer: ```suggestion
In the -stable branches (contrary to main branch), the release crew can directly push new commits to quickly address issues. After applying the necessary changes, those changes need to be re-implemented in `main`; this is to avoid inconsistent state across branches, and to prevent having the same failures in the future releases.
```

To make sure not to forget about these changes, we follow some best practices.

1. When adding a local commit to the `stable` branch, the commit message should follow the format:
   `[LOCAL] <fix or change applied to the repo>`
2. After the commit:
   1. open the **Road to 0.X.0** discussion (e.g: [Road to 0.71.0](https://github.com/reactwg/react-native-releases/discussions/41)) or the **Should we release 0.X.Y** discussion (e.g.: [Should we release 0.70.7](https://github.com/reactwg/react-native-releases/discussions/48))
   2. edit the main message adding the link to your commit under the **Local commits to backport to main** section

> Right after the release, make sure to backport those commits into main. When the backport PR is open against `main`, make sure to add it right below the commit that has to be backported. In this way, we know that that commit has been taken care of.

Reviewer: ```suggestion
Once the issue in the stable branch is handled, make sure to backport the fixes to main; when the PR backporting the fix is open against `main`, make sure to update the relevant release discussion by adding the link to this PR right below the local commit in the **Local commits to backport to main** section. This makes it easier to keep track of the fact that the local commit has been taken care of.
```
